### PR TITLE
Converted PcaTransform into Transformer using TransformerWrapper.

### PR DESCRIPTION
**Pull request by zeahmed**
This PR implements a work item related to #754.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/WrappedPcaTransform.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        /// <param name="weightColumn">The name of the weight column.</param>
+        /// <param name="rank">The number of components in the PCA.</param>
+        /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
+        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="seed">The seed for random number generation.</param>
+        public PcaEstimator(IHostEnvironment env,
+            string inputColumn,
+            string outputColumn = null,
+            string weightColumn = null,
+            int rank = 20,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/WrappedPcaTransform.cs]: >20 [](start = 23, length = 2)

Defaults class please.

File: test/Microsoft.ML.Tests/Transformers/PcaTests.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                .Read(new MultiFileSource(dataSource));
+
+            var invalidData = TextLoader.CreateReader(env,
+                c => (label: c.LoadFloat(11), features: c.LoadText(0, 10)),
+                separator: ';', hasHeader: true)
+                .Read(new MultiFileSource(dataSource));
+
+            var est = new PcaEstimator(env, "features", "pca", null, rank: 5, seed: 1);
+            //TestEstimatorCore(est, data.AsDynamic, invalidInput: invalidData.AsDynamic);
+
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Transformers/PcaTests.cs]: ?

Comment (zeahmed) [reply test/Microsoft.ML.Tests/Transformers/PcaTests.cs]: failing on empty data view. Added description and link to issue.

---
In reply to: [220037180](https://github.com/dotnet/machinelearning/pull/1017#discussion_r220037180) [](ancestors = 220037180)

File: src/Microsoft.ML.Transforms/WrappedPcaTransform.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        /// <param name="rank">The number of components in the PCA.</param>
+        /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
+        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="seed">The seed for random number generation.</param>
+        public static Vector<float> ToPrincipalComponents(this Vector<float> input,
+            Vector<float> weightColumn,
+            int rank = 20,
+            int oversampling = 20,
+            bool center = true,
+            int seed = 0) => new OutPipelineColumnWithWeights(input, weightColumn, rank, oversampling, center, seed);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/WrappedPcaTransform.cs]: >weightColumn, rank, oversampling, center, seed [](start = 69, length = 46)

I would prefer Config class/structure as we have in Categorical/Term/NAReplace than passing multiple params

File: src/Microsoft.ML.PCA/PcaTransform.cs [hunk 4]
```diff
@@ -29,25 +29,33 @@ namespace Microsoft.ML.Runtime.Data
     /// <include file='doc.xml' path='doc/members/member[@name="PCA"]/*' />
     public sealed class PcaTransform : OneToOneTransformBase
     {
+        public static class Defaults
```
Comment (TomFinley) [inline src/Microsoft.ML.PCA/PcaTransform.cs]: >Defaults [](start = 28, length = 8)

Is there a particular reason to make this public?

File: src/Microsoft.ML.Transforms/WrappedPcaTransform.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.ML.Transforms
+{
+    /// <include file='doc.xml' path='doc/members/member[@name="PCA"]/*' />
+    public sealed class PcaEstimator : TrainedWrapperEstimatorBase
```
Comment (TomFinley) [inline src/Microsoft.ML.Transforms/WrappedPcaTransform.cs]: >PcaEstimator [](start = 24, length = 12)

This should not be here. This should be in the same assembly as PCA.

File: src/Microsoft.ML.Transforms/Microsoft.ML.Transforms.csproj [hunk 6]
```diff
@@ -48,6 +48,7 @@
     <ProjectReference Include="..\Microsoft.ML.Api\Microsoft.ML.Api.csproj" />
     <ProjectReference Include="..\Microsoft.ML.Core\Microsoft.ML.Core.csproj" />
     <ProjectReference Include="..\Microsoft.ML.Data\Microsoft.ML.Data.csproj" />
+    <ProjectReference Include="..\Microsoft.ML.PCA\Microsoft.ML.PCA.csproj" />
```
Comment (TomFinley) [inline src/Microsoft.ML.Transforms/Microsoft.ML.Transforms.csproj]: ><ProjectReference Include="..\Microsoft.ML.PCA\Microsoft.ML.PCA.csproj" /> [](start = 4, length = 74)

Once you move the file to the correct location, please remember to remove this reference.

File: src/Microsoft.ML.PCA/PcaTransform.cs (hunk 4 above)
Comment (zeahmed) [reply src/Microsoft.ML.PCA/PcaTransform.cs]: After moving the file into the same project, I am making it `internal`. I am using it in `PcaEstimator` class for managing defaults.

---
In reply to: [220046677](https://github.com/dotnet/machinelearning/pull/1017#discussion_r220046677) [](ancestors = 220046677)

File: src/Microsoft.ML.PCA/PcaTransform.cs [hunk 7]
```diff
@@ -29,25 +29,33 @@ namespace Microsoft.ML.Runtime.Data
...
+        {
+            public const string WeightColumn = null;
+            public const int Rank = 20;
+            public const int Oversampling = 20;
+            public const bool Center = true;
+            public const int Seed = 0;
+        }
         public sealed class Arguments : TransformInputBase
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.PCA/PcaTransform.cs]: it so cramped here, if only we have new line to separate them!

**Pull request merged**